### PR TITLE
Role management

### PR DESCRIPTION
**Pull request by olive42**

### CODE REVIEW

File: attributes/api.rb [hunk 1]
```diff
@@ -2,6 +2,8 @@
 default['nexus3']['api']['type'] = 'groovy'
 default['nexus3']['api']['host'] = 'http://localhost:8081'
 default['nexus3']['api']['endpoint'] = "#{node['nexus3']['api']['host']}/service/siesta/rest/v1/script/"
+default['nexus3']['api']['username'] = 'admin'.freeze
```
[the remark this answers is not in the document]
Comment (olive42) [reply attributes/api.rb]: This was done in b23b15d. I rebased on master, let's see what Travis thinks.

File: resources/role.rb [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+property :roles, Array, default: lazy { [] }
+property :privileges, Array, default: lazy { [] }
+property :api_endpoint, String, identity: true, default: lazy { node['nexus3']['api']['endpoint'] }
+property :api_username, String, identity: true, default: lazy { node['nexus3']['api']['username'] }
+property :api_password, String, identity: true, sensitive: true, default: lazy { node['nexus3']['api']['password'] }
+
+load_current_value do |desired|
+  api_endpoint desired.api_endpoint
+  api_username desired.api_username
+  api_password desired.api_password
```
Comment (Annih) [inline resources/role.rb]: The 3 above lines are not needed because these properties are "identity"

File: resources/role.rb [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+  begin
+    res = ::Nexus3::Api.new(api_endpoint, api_username, api_password).run_script('get_role', desired.role_name)
```
Comment (Annih) [inline resources/role.rb]: you can use `role_name` instead of `desired.role_name`

File: resources/role.rb [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    current_value_does_not_exist! if res == 'null'
+    config = JSON.parse(res)
```
Comment (Annih) [inline resources/role.rb]: OPT: you can directly parse the result L15 and compare with nil L16

File: resources/role.rb [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    current_value_does_not_exist! if config.nil?
```
Comment (Annih) [inline resources/role.rb]: This is unlikely to happen because you already filtered the `null` result

File: resources/role.rb [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    ::Chef::Log.debug "Role config is #{config}"
+    role_name config['role']
```
Comment (Annih) [inline resources/role.rb]: Useless because this is the "name property"

File: resources/role.rb [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    description config['description']
+    roles config['roles'].sort!
+    privileges config['privileges'].sort!
+  rescue LoadError, ::Nexus3::ApiError => e
```
Comment (Annih) [inline resources/role.rb]: OPT: `::LoadError`

File: resources/role.rb [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+  end
+end
+
+action_class do
+  def init
+    chef_gem 'httpclient' do
+      compile_time true
+    end
+
+    nexus3_api "get_role #{new_resource.role_name}" do
```
Comment (Annih) [inline resources/role.rb]: Why do you use the `role_name` here?
You just need to create the script not to execute it.

File: resources/role.rb [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+      action :create
+      script_name 'get_role'
```
Comment (Annih) [inline resources/role.rb]: Useless if you set a good name to the resource

File: resources/role.rb [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+      endpoint new_resource.api_endpoint
+      username new_resource.api_username
+      password new_resource.api_password
+      args new_resource.role_name
```
Comment (Annih) [inline resources/role.rb]: Useless here, you just need to create the script

File: resources/role.rb [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+authManager = security.getSecuritySystem().getAuthorizationManager(UserManager.DEFAULT_SOURCE);
+try {
+    existingRole = authManager.getRole(args);
+    log.info("Deleting role ${existingRole.getName()}");
+    authManager.deleteRole(existingRole.getRoleId());
+} catch (NoSuchRoleException ignored) {
+    // No such role, ignoring.
+}
+    EOS
```
Comment (Annih) [inline resources/role.rb]: LINT: indent 2 spaces more

File: resources/role.rb [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+  begin
+    res = ::Nexus3::Api.new(api_endpoint, api_username, api_password).run_script('get_role', desired.role_name)
+    current_value_does_not_exist! if res == 'null'
+    config = JSON.parse(res)
+    current_value_does_not_exist! if config.nil?
+    ::Chef::Log.debug "Role config is #{config}"
+    role_name config['role']
+    description config['description']
+    roles config['roles'].sort!
+    privileges config['privileges'].sort!
```
Comment (Annih) [inline resources/role.rb]: The script is already sorting, do you need to do it twice?

File: resources/role.rb (hunk 2 above)
Comment (olive42) [reply resources/role.rb]: Done

File: resources/role.rb (hunk 3 above)
Comment (olive42) [reply resources/role.rb]: done

File: resources/role.rb (hunk 4 above)
Comment (olive42) [reply resources/role.rb]: done

File: resources/role.rb (hunk 5 above)
Comment (olive42) [reply resources/role.rb]: done

File: resources/role.rb (hunk 6 above)
Comment (olive42) [reply resources/role.rb]: done

File: resources/role.rb (hunk 12 above)
Comment (olive42) [reply resources/role.rb]: One can never be too sure about what's returned.

File: resources/role.rb (hunk 7 above)
Comment (olive42) [reply resources/role.rb]: done

File: resources/role.rb (hunk 11 above)
Comment (olive42) [reply resources/role.rb]: done

File: resources/role.rb (hunk 8 above)
Comment (olive42) [reply resources/role.rb]: Removed.

File: resources/role.rb (hunk 10 above)
Comment (olive42) [reply resources/role.rb]: done

**Pull request merged**